temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowService.java

```diff
@@ -777,14 +782,30 @@ public void respondNexusTaskCompleted(
       RespondNexusTaskCompletedRequest request,
       StreamObserver<RespondNexusTaskCompletedResponse> responseObserver) {
     try {
-      NexusTaskToken taskToken = NexusTaskToken.fromBytes(request.getTaskToken());
-      TestWorkflowMutableState mutableState = getMutableState(taskToken.getExecutionId());
-      if (request.getResponse().hasStartOperation()
-          && request.getResponse().getStartOperation().hasAsyncSuccess()) {
-        // Start event is only recorded for async success
-        mutableState.startNexusTask(taskToken.getScheduledEventId(), request);
+      NexusOperationRef ref = NexusOperationRef.fromBytes(request.getTaskToken());
+      TestWorkflowMutableState mutableState = getMutableState(ref.getExecutionId());
+      if (request.getResponse().hasCancelOperation()) {
+        mutableState.cancelNexusOperation(ref);
+      } else if (request.getResponse().hasStartOperation()) {
+        StartOperationResponse startResp = request.getResponse().getStartOperation();
+        if (startResp.hasOperationError()) {
+          Failure failure = unsuccessfulOperationErrorToFailure(startResp.getOperationError());
+          mutableState.failNexusOperation(ref, failure);
+        } else if (startResp.hasAsyncSuccess()) {
+          // Start event is only recorded for async success
+          mutableState.startNexusOperation(
+              ref.getScheduledEventId(), request.getIdentity(), startResp.getAsyncSuccess());
+        } else if (startResp.hasSyncSuccess()) {
+          mutableState.completeNexusOperation(ref, startResp.getSyncSuccess().getPayload());
```

Comment:
Here you want to use the attempts and cancel state to verify the task token (which you do, just clarifying that this is different from async completion).